src/spetlr/power_bi/PowerBi.py

```diff
@@ -0,0 +1,436 @@
+        self.last_exception = None
+        self.last_refresh_utc = None
+        self.last_duration = 0
+
+    def _raise_error(self, message: str):
+        if self.ignore_errors:
+            print(message)
+        else:
+            raise Exception(message)
+
+    def _raise_api_error(self, message: str, api_call: requests.Response):
+        print(api_call.text)
+        self._raise_error(
+            message + f" Response: {api_call.status_code} {api_call.reason}"
+        )
+
+    def _get_access_token(self) -> bool:
+        """
+        Acquires an access token to connect to PowerBI.
+
+        :return: True if succeeded or False if failed (when ignore_errors==True)
+        :rtype: bool
+        :raises Exception: if failed and ignore_errors==False
+        """
+
+        if self.expire_time != 0:
+            if time.time() < self.expire_time:
+                return True
+            print("Renewing access token...")
+
+        # Prepare URLs (note: we fetch only the latest refresh record, i.e. top=1)
+        authority_url = f"https://login.microsoftonline.com/{self.client.tenant_id}/"
+        scope = ["https://analysis.windows.net/powerbi/api/.default"]
+        self.powerbi_url = "https://api.powerbi.com/v1.0/myorg/"
+
+        # Use MSAL to get an access token
+        app = msal.ConfidentialClientApplication(
+            self.client.client_id,
+            authority=authority_url,
+            client_credential=self.client.client_secret,
+        )
+        result = app.acquire_token_for_client(scopes=scope)
+        if "access_token" not in result:
+            print(result)
+            self._raise_error("Failed to acquire token!")
+            return False
+
+        access_token = result["access_token"]
+        if "expires_in" in result:
+            self.expire_time = time.time() + result["expires_in"] - 3
+        else:
+            self.expire_time = time.time() + (5 * 60) - 3
+
+        # Get latest Power BI Dataset refresh record
+        self.api_header = {
+            "Content-Type": "application/json",
+            "Authorization": f"Bearer {access_token}",
+        }
+        return True
+
+    def _get_workspace(self) -> bool:
+        """
+        Gets workspace ID based on the workspace name, or shows all workspaces
+            if no parameter was specified.
+
+        :return: True if succeeded or False if failed (when ignore_errors==True)
+        :rtype: bool
+        :raises Exception: if failed and ignore_errors==False
+        """
+
+        if self.workspace_id is None:
+            api_call = requests.get(
+                url=f"{self.powerbi_url}groups", headers=self.api_header
+            )
+            if api_call.status_code != 200:
+                self._raise_api_error("Failed to fetch workspaces!", api_call)
+                return False
+            df = pd.DataFrame(api_call.json()["value"], columns=["id", "name"])
+
+            if self.workspace_name is None:
```

Comment:
I think this display of workspaces should be in its own function.

Reply:
Ok